source/s2n/s2n_tls_channel_handler.c

```diff
@@ -325,14 +327,34 @@ static int s_s2n_handler_send(void *io_context, const uint8_t *buf, uint32_t len
 static void s_s2n_handler_destroy(struct aws_channel_handler *handler) {
     if (handler) {
         struct s2n_handler *s2n_handler = (struct s2n_handler *)handler->impl;
+        aws_crt_statistics_tls_cleanup(&s2n_handler->stats);
         s2n_connection_free(s2n_handler->connection);
         aws_mem_release(handler->alloc, (void *)s2n_handler);
     }
 }
 
+static void s_on_negotiation_result(
+    struct aws_channel_handler *handler,
+    struct aws_channel_slot *slot,
+    int error_code,
+    void *user_data) {
+
+    struct s2n_handler *s2n_handler = (struct s2n_handler *)handler->impl;
+    if (s2n_handler->on_negotiation_result) {
+        s2n_handler->on_negotiation_result(handler, slot, error_code, user_data);
+    }
+
+    s2n_handler->stats.handshake_status =
+        (error_code == AWS_ERROR_SUCCESS) ? AWS_MTLS_STATUS_SUCCESS : AWS_MTLS_STATUS_FAILURE;
+}
+
 static int s_drive_negotiation(struct aws_channel_handler *handler) {
     struct s2n_handler *s2n_handler = (struct s2n_handler *)handler->impl;
 
+    if (s2n_handler->stats.handshake_status == AWS_MTLS_STATUS_NONE) {
```

Comment:
Same here, should we assert on this? Could it be ```SUCCESS```/```ERROR```, and then we're trying again?

Reply:
Reworked and added asserts